This folder contains notebooks and scripts related to all the single-cell and spatial analysis of this paper.

## Contents

1) `01_preprocessing_singlecell_and_Visium`: scripts to preprocess and quality control of single-cell and spatial transcriptomics.
2) `02_number_of_cells_and_spots`: Number of cells and spots that passed QC per donor.
3) `03_single_cell_annotation`: Dotplots of markers and visualisations for single-cell annotation.
4) `04_single_cell_differential_expression`: scripts for differential gene expression using GLM.
5) `05_single_cell_differential_communication`: scripts for single-cell differential cell-to-cell communication. Selected cell types.
6) `06_single_cell_differential_proportions`: scripts and visualisations for differential proportions.
7) `07_single_cell_pseudotime`: brief description of pseudotime analysis. Link to a detail github repository.
8) `08_single_cell_DA_overlap`: scripts to calculate differential abundance overlap between gestational age and condition.
9) `09_spatial_score`: scripts to calculate an hypoxia spatial score per capture area. 
10) `10_spatial_deconvolution`: scripts to deconvolute Visium spatial transcriptomics.
11) `11_spatial_pattern_extraplacental_trophoblasts`: scripts and notebooks related to the spatial pattern of trophoblasts at the FMI.
12) `12_neighbourhoods_spatial_domains`: notebook to calculate the most frequent spatial domain.  

